lib/SPIRV/SPIRVReader.cpp

```diff
@@ -550,24 +550,69 @@ void SPIRVToLLVM::setName(llvm::Value *V, SPIRVValue *BV) {
 void SPIRVToLLVM::setLLVMLoopMetadata(SPIRVLoopMerge *LM, BranchInst *BI) {
   if (!LM)
     return;
-  llvm::MDString *Name = nullptr;
   auto Temp = MDNode::getTemporary(*Context, None);
   auto Self = MDNode::get(*Context, Temp.get());
   Self->replaceOperandWith(0, Self);
-
   if (LM->getLoopControl() == LoopControlMaskNone) {
     BI->setMetadata("llvm.loop", Self);
     return;
-  } else if (LM->getLoopControl() == LoopControlUnrollMask)
+  }
+
+  llvm::MDString *Name;
+  std::vector<SPIRVWord> LoopControlParameters;
+  if (LM->getLoopControl() & LoopControlUnrollMask) {
     Name = llvm::MDString::get(*Context, "llvm.loop.unroll.full");
-  else if (LM->getLoopControl() == LoopControlDontUnrollMask)
+    generateLLVMLoopMetadata(BI, Name, /*LoopControl parameter*/ 0);
+  }
+  else if (LM->getLoopControl() & LoopControlDontUnrollMask) {
     Name = llvm::MDString::get(*Context, "llvm.loop.unroll.disable");
-  else
-    return;
+    generateLLVMLoopMetadata(BI, Name, /*LoopControl parameter*/ 0);
+  }
+  if (LM->getLoopControl() & LoopControlDependencyInfiniteMask) {
+    Name = llvm::MDString::get(*Context, "llvm.loop.ivdep.enable");
+    generateLLVMLoopMetadata(BI, Name, /*LoopControl parameter*/ 0);
+  }
+  if (LM->getLoopControl() & LoopControlDependencyLengthMask) {
+    LoopControlParameters = LM->getLoopControlParameters();
+    Name = llvm::MDString::get(*Context, "llvm.loop.ivdep.safelen");
+    if (!LoopControlParameters.empty())
+      generateLLVMLoopMetadata(BI, Name, LoopControlParameters[0]);
+  }
+  if (LM->getLoopControl() & LoopControlExtendedControlsMask) {
+    LoopControlParameters = LM->getLoopControlParameters();
+    for (auto LCP = LoopControlParameters.begin();
+         LCP != LoopControlParameters.end(); ++LCP) {
+      switch (*LCP) {
+      case InitiationIntervalINTEL: {
+        Name = llvm::MDString::get(*Context, "llvm.loop.ii.count");
+        generateLLVMLoopMetadata(BI, Name, *(++LCP));
+        break;
+      }
+      case MaxConcurrencyLoopINTEL: {
+        Name = llvm::MDString::get(*Context, "llvm.loop.max_concurrency.count");
+        generateLLVMLoopMetadata(BI, Name, *(++LCP));
+        break;
+      }
+      default:
+        break;
+      }
+    }
+  }
+}
 
+void SPIRVToLLVM::generateLLVMLoopMetadata(BranchInst *BI,
+                                           llvm::MDString *Name,
+                                           SPIRVWord Parameter) {
+  auto Temp = MDNode::getTemporary(*Context, None);
```

Comment:
Please add a comment explaining what's going on in this function.

Reply:
Removed function